core/domain/state_domain.py

```diff
@@ -1279,6 +1279,19 @@ def get_translation_counts(self):
 
         return translation_counts
 
+    def get_translation_content_count(self):
+        """Returns the number of content IDs that map to non-empty content
+        objects.
+
+        Returns:
+            int. The number of content IDs mapping to non-empty content objects.
+        """
+        count = 0
+        for translations in self.translations_mapping.values():
+            if translations:
```

Comment:
Here translation will represent the language_code to written_translation_dict, right? So it's translations is `{}` then it represent that the given content-id doesn't have translations in any language, right? is that what we are doing here? [I'm unable to understand the value returned by this method, can you please explain?]

Reply:
Yes, since currently the denominator for translation opportunity progress is determined by exploration `content_count`, which downstream counts all written translations. An alternative approach would be to make sure each of the content IDs map to a written translation with the `needs_update` field set to false to account for the following case:

```
written_translations:
  translations_mapping:
    content: {hi: {…}}
    default_outcome: {}
```

Since the `translation_counts` used for the numerator of the opportunity progress depends on the written translation having the `needs_update` field:

```
translation_counts = collections.defaultdict(int)
for translations in self.translations_mapping.values():
    for language, translation in translations.items():
        if not translation.needs_update:
            translation_counts[language] += 1
return translation_counts
```

Reply:
@DubeySandeep Can you also point me to where the `translations_mapping` gets populated? I'm wondering why loading dummy explorations in the dev server populates some `{language_code -> written_translation}` dicts but not others.

Reply:
> Can you also point me to where the translations_mapping gets populated?

We have a change command which completely changes the written_translation property of the state. (Though we have add_translation command which will be used in translation suggestion.)

`content_id` will have dict with the language code key for only the languages in which translation available! So for dummy exploration maybe we only have translations for a few content_ids in a few languages?

Reply:
Hmm, but it doesn't seem like the translation mapping gets initialized even when the translation is available, e.g. consider the following exploration state:
<img width="979" alt="Screen Shot 2020-08-06 at 1 40 11 PM" src="https://user-images.githubusercontent.com/9004726/89564021-62e95780-d7ea-11ea-9c18-166669f67d87.png">

The translation mapping is empty, even though there's content available for translation (the top level content html: "Congratulations, you have finished!").

Reply:
translation_mapping.content is empty because it doesn't have any translations for the HTML with id `content`.

Assume you add a Hindi translation (translation: बधाई हो, आपने समाप्त कर दिया है!)for the  HTML with content_id `content`, in that case the written translation propert will be changed to:

```
written_translations: {
    translations_mapping: {
         hi: {
             html: 'बधाई हो, आपने समाप्त कर दिया है!',
             needs_update: False
         }
     }
}
```